opentelemetry-sdk/src/testing/span/in_memory_exporter.rs

```diff
@@ -0,0 +1,150 @@
+///# use opentelemetry_sdk::runtime;
+///# use opentelemetry_sdk::testing::span::InMemorySpanExporterBuilder;
+///# use opentelemetry_sdk::trace::{BatchSpanProcessor, TracerProvider};
+///
+///# #[tokio::main]
+///# async fn main() {
+///     global::set_text_map_propagator(TraceContextPropagator::new());
+///     let exporter = InMemorySpanExporterBuilder::new().build();
+///     let provider = TracerProvider::builder()
+///         .with_span_processor(BatchSpanProcessor::builder(exporter.clone(), runtime::Tokio).build())
+///         .build();
+///
+///     global::set_tracer_provider(provider.clone());
+///
+///     let tracer = global::tracer("example/in_memory_exporter");
+///     let span = tracer
+///         .span_builder("say hello")
+///         .with_kind(SpanKind::Server)
+///         .start(&tracer);
+///
+///     let cx = Context::current_with_span(span);
+///     cx.span().add_event("handling this...", Vec::new());
+///     cx.span().end();
+///
+///     let results = provider.force_flush();
+///     for result in results {
+///         if let Err(e) = result {
+///             println!("{:?}", e)
+///         }
+///     }
+///     let spans = exporter.get_finished_spans().unwrap();
+///     for span in spans {
+///         println!("{:?}", span)
+///     }
+///# }
+/// ```
+#[derive(Clone, Debug)]
+pub struct InMemorySpanExporter {
+    spans: Arc<Mutex<Vec<SpanData>>>,
+}
+
+impl Default for InMemorySpanExporter {
+    fn default() -> Self {
+        InMemorySpanExporterBuilder::new().build()
+    }
+}
+
+/// Builder for [`InMemorySpanExporter`].
+/// # Example
+/// ```
+///# use opentelemetry_sdk::testing::span::InMemorySpanExporterBuilder;
+///
+/// let exporter = InMemorySpanExporterBuilder::new().build();
+/// ```
+#[derive(Clone, Debug)]
+pub struct InMemorySpanExporterBuilder {}
+
+impl Default for InMemorySpanExporterBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl InMemorySpanExporterBuilder {
+    /// Creates a new instance of the `InMemorySpanExporterBuilder`.
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    /// Creates a new instance of the `InMemorySpanExporter`.
+    pub fn build(&self) -> InMemorySpanExporter {
+        InMemorySpanExporter {
+            spans: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+}
+
+impl InMemorySpanExporter {
+    /// Returns the finished span as a vector of `SpanData`.
+    ///
+    /// # Errors
+    ///
+    /// Returns a `TraceError` if the internal lock cannot be acquired.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use opentelemetry_sdk::testing::span::InMemorySpanExporter;
+    ///
+    /// let exporter = InMemorySpanExporter::default();
+    /// let finished_spans = exporter.get_finished_spans().unwrap();
+    /// ```
+    pub fn get_finished_spans(&self) -> TraceResult<Vec<SpanData>> {
+        self.spans
+            .lock()
+            .map(|spans_guard| spans_guard.iter().map(Self::clone_span).collect())
```

Comment:
Check if you can use `iter().cloned().collect()` 
ref:
[comment](https://github.com/open-telemetry/opentelemetry-rust/pull/1231#discussion_r1307589254)
[my commit](https://github.com/open-telemetry/opentelemetry-rust/pull/1231/commits/03303122f5aaeb6603689f35332aab6dd217a945)